### PR TITLE
[RLlib] Schedule-classes multi-framework support.

### PR DESCRIPTION
**Pull request by sven1977**
Currently, our RLlib Schedule classes only support pure python. They do not work in a tf/torch setting. tf/torch support is required for the currently developed RLlib Exploration API (i.e. requires Schedules to work in-graph).
 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: rllib/utils/framework.py [hunk 1]
```diff
@@ -41,12 +69,17 @@ def try_import_tfp():
...
         return tfp
-    except ImportError:
+    except ImportError as e:
+        if error:
+            raise e
         return None
 
 
-def try_import_torch():
+def try_import_torch(error=False):
```
Comment (ericl) [inline rllib/utils/framework.py]: this isn't implemented

File: rllib/utils/schedules/polynomial_schedule.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        self.initial_p = initial_p
+        self.power = power
+
+    def value(self, t):
+        """
+        Returns the result of:
+        final_p + (initial_p - final_p) * (1 - `t`/t_max) ** power
+        """
+        if self.framework == "tf" and tf.executing_eagerly() is False:
+            return tf.train.polynomial_decay(
```
Comment (ericl) [inline rllib/utils/schedules/polynomial_schedule.py]: I'm wondering if schedules should support frameworks. We could always call them in Python land and feed their values in, which would make them framework independent automatically (as they are currently I believe).

Comment (ericl) [reply rllib/utils/schedules/polynomial_schedule.py]: Basically, instead of feeding in a `t` tensor and letting the schedule be computed in-graph we feed in the `value` tensor (after computing the schedule out of graph).

Comment (sven1977) [reply rllib/utils/schedules/polynomial_schedule.py]: The problem is when stuff happens deeper inside some components. How is Policy supposed to know that its Exploration object (which could be of any type) needs some Schedule's value (or even know that the Exploration object has a Schedule (not all Explorations do))?

File: rllib/utils/framework.py (hunk 1 above)
Comment (sven1977) [reply rllib/utils/framework.py]: Taking a look.

Comment (sven1977) [reply rllib/utils/framework.py]: Done.

File: rllib/utils/schedules/polynomial_schedule.py (hunk 2 above)
Comment (sven1977) [reply rllib/utils/schedules/polynomial_schedule.py]: An example, where a Schedule could be really buried in the middle of the graph: In SAC discrete, if you were to use a decaying temperature parameter for your Gumbel Softmax distribution, the sampling forward pass through the Gumbel is right in the middle between two NN forward passes (the policy NN to sample, then the Q-nets to get the action-values), which together as a whole have to be differentiable. Sure, you could still get the current decayed weight value and then pass it in, but it would make everything much more complicated for the user.

Comment (ericl) [reply rllib/utils/schedules/polynomial_schedule.py]: Ok, I think there is a way we could allow Schedules to be embedded in graphs without having to do too much work to support different frameworks.

Since we don't need to differentiate through schedules, and they are not performance critical, we could add `.tf_value(time: Tensor) -> Tensor` and `.torch_value(time: Tensor) -> Tensor` helper methods.

You can implement this straightforwardly in torch / tf eager. In TF graph mode you can use `tf.py_function` (which isn't differentiable, but can allow arbitrary python code to be run in the graph). How does this sound?

Comment (ericl) [reply rllib/utils/schedules/polynomial_schedule.py]: (this would also be similar to how the Exploration interface works with diff frameworks I believe).

**Review by ericl — APPROVED** [review]

File: rllib/utils/schedules/schedule.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    Useful for backend-agnostic rate/weight changes for learning rates,
+    exploration epsilons, beta parameters for prioritized replay, loss weights
+    decay, etc..
+
+    Each schedule can be called directly with the `t` (absolute time step)
+    value and returns the value dependent on the Schedule and the passed time.
+    """
+
+    def __init__(self, framework=None):
+        # TODO(sven): replace with .tf_value() / torch_value() methods that
```
Comment (ericl) [inline rllib/utils/schedules/schedule.py]: Going to merge this, we should address this TODO in a followup PR.
